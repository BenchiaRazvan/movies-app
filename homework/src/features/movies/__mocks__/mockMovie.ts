export const mockMovie = {
    Title: "The Beta Test",
    Year: "2021",
    imdbID: "tt11738830",
    Type: "movie",
    Poster:
      "https://m.media-amazon.com/images/M/MV5BZTBhNjcyMjEtM2NiMi00YzQ3LThlYjEtMjFiNDc3NWJiNGJiXkEyXkFqcGc@._V1_SX300.jpg",
  };