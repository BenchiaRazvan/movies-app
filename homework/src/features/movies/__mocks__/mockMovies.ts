export const mockMovies = [
  {
    Title: "The Test Case",
    Year: "2018–",
    imdbID: "tt6868278",
    Type: "series",
    Poster:
      "https://m.media-amazon.com/images/M/MV5BZjNjOTQzOWItMzk4MC00MmMzLTk4ZjUtMWU4ZWE4ZmU0OWY5XkEyXkFqcGc@._V1_SX300.jpg",
  },
  {
    Title: "Beta Test",
    Year: "2016",
    imdbID: "tt4244162",
    Type: "movie",
    Poster:
      "https://m.media-amazon.com/images/M/MV5BZjhlM2ZhMzUtMzU4ZC00ZjAyLWIxZmYtOWY4N2RlMWEzYTcwXkEyXkFqcGc@._V1_SX300.jpg",
  },
  {
    Title: "Test",
    Year: "2013",
    imdbID: "tt2407380",
    Type: "movie",
    Poster:
      "https://m.media-amazon.com/images/M/MV5BMTQwMDU5NDkxNF5BMl5BanBnXkFtZTcwMjk5OTk4OQ@@._V1_SX300.jpg",
  },
  {
    Title: "The Test",
    Year: "2012",
    imdbID: "tt1986180",
    Type: "movie",
    Poster:
      "https://m.media-amazon.com/images/M/MV5BMTYwNTgzMjM5M15BMl5BanBnXkFtZTcwNDUzMTE1OA@@._V1_SX300.jpg",
  },
  {
    Title: "Baka and Test",
    Year: "2010–2011",
    imdbID: "tt1655610",
    Type: "series",
    Poster:
      "https://m.media-amazon.com/images/M/MV5BOGRmOTAwNDQtYzQ5NS00YmZjLThkMDItNzYyMmVlYjIxZmZjXkEyXkFqcGc@._V1_SX300.jpg",
  },
];
