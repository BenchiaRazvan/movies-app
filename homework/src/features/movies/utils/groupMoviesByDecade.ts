import { MAXIMUM_NUMBER_OF_DECADES } from "../../../shared/constants/numberConstants";
import { Movie } from "../types/movieTypes";

export const parseMovieYear = (year: string): number => {
  const parsedYear = year.includes("–")
    ? parseInt(year.split("–")[0])
    : parseInt(year);

  return parsedYear;
};

export const groupMoviesByDecade = (movies: Movie[]) => {
  movies.sort((currentMovie, nextMovie) => {
    const currentMovieYear = parseMovieYear(currentMovie.Year);
    const nextMovieYear = parseMovieYear(nextMovie.Year);

    return nextMovieYear - currentMovieYear;
  });

  let numberOfDecades = 0;

  const groupedMovies = movies.reduce((acc: Map<string, Movie[]>, movie) => {
    const movieYear = parseMovieYear(movie.Year);
    const decadeYear = (movieYear - (movieYear % 10)).toString();

    if (!acc.has(decadeYear) && numberOfDecades < MAXIMUM_NUMBER_OF_DECADES) {
      acc.set(decadeYear, []);
      numberOfDecades++;
    }

    acc.get(decadeYear)?.push(movie);

    return acc;
  }, new Map());

  return groupedMovies;
};
