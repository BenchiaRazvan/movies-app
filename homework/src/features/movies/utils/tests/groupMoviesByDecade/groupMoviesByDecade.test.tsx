import "@testing-library/jest-dom";
import { groupMoviesByDecade, parseMovieYear } from "../../groupMoviesByDecade";
import { mockMovies } from "../../../__mocks__/mockMovies";

describe("CardComponent", () => {
  it("parse move year to number", () => {
    expect(parseMovieYear("1990")).toBe(1990);
    expect(parseMovieYear("2022-2205")).toBe(2022);
  });

  it("group movies by decades", () => {
    const result = groupMoviesByDecade(mockMovies);
    expect(result).toMatchSnapshot();
  });
});
