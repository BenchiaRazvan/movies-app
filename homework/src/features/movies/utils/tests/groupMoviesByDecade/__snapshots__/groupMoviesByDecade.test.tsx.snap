// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardComponent group movies by decades 1`] = `
Map {
  "2010" => Array [
    Object {
      "Poster": "https://m.media-amazon.com/images/M/MV5BZjNjOTQzOWItMzk4MC00MmMzLTk4ZjUtMWU4ZWE4ZmU0OWY5XkEyXkFqcGc@._V1_SX300.jpg",
      "Title": "The Test Case",
      "Type": "series",
      "Year": "2018–",
      "imdbID": "tt6868278",
    },
    Object {
      "Poster": "https://m.media-amazon.com/images/M/MV5BZjhlM2ZhMzUtMzU4ZC00ZjAyLWIxZmYtOWY4N2RlMWEzYTcwXkEyXkFqcGc@._V1_SX300.jpg",
      "Title": "Beta Test",
      "Type": "movie",
      "Year": "2016",
      "imdbID": "tt4244162",
    },
    Object {
      "Poster": "https://m.media-amazon.com/images/M/MV5BMTQwMDU5NDkxNF5BMl5BanBnXkFtZTcwMjk5OTk4OQ@@._V1_SX300.jpg",
      "Title": "Test",
      "Type": "movie",
      "Year": "2013",
      "imdbID": "tt2407380",
    },
    Object {
      "Poster": "https://m.media-amazon.com/images/M/MV5BMTYwNTgzMjM5M15BMl5BanBnXkFtZTcwNDUzMTE1OA@@._V1_SX300.jpg",
      "Title": "The Test",
      "Type": "movie",
      "Year": "2012",
      "imdbID": "tt1986180",
    },
    Object {
      "Poster": "https://m.media-amazon.com/images/M/MV5BOGRmOTAwNDQtYzQ5NS00YmZjLThkMDItNzYyMmVlYjIxZmZjXkEyXkFqcGc@._V1_SX300.jpg",
      "Title": "Baka and Test",
      "Type": "series",
      "Year": "2010–2011",
      "imdbID": "tt1655610",
    },
  ],
}
`;
